crates/matrix-sdk/src/widget/mod.rs

```diff
@@ -141,89 +135,132 @@ impl WidgetDriver {
+        for action in initial_actions {
+            ctx.process_action(action).await?;
+        }
 
-                        MatrixDriverRequestData::GetOpenId => matrix_driver
-                            .get_open_id()
-                            .await
-                            .map(MatrixDriverResponse::OpenIdReceived)
-                            .map_err(|e| e.to_string()),
+        // Process incoming events.
+        while let Some(event) = events_rx.recv().await {
+            ctx.process_event(event).await?;
+        }
 
-                        MatrixDriverRequestData::ReadMessageLikeEvent(cmd) => matrix_driver
-                            .read_message_like_events(cmd.event_type.clone(), cmd.limit)
-                            .await
-                            .map(MatrixDriverResponse::MatrixEventRead)
-                            .map_err(|e| e.to_string()),
+        Ok(())
+    }
+}
 
-                        MatrixDriverRequestData::ReadStateEvent(cmd) => matrix_driver
-                            .read_state_events(cmd.event_type.clone(), &cmd.state_key)
-                            .await
-                            .map(MatrixDriverResponse::MatrixEventRead)
-                            .map_err(|e| e.to_string()),
+/// A small wrapper of all the data that we need to process an incoming event.
+struct ProcessingContext<T> {
+    widget_machine: WidgetMachine,
+    matrix_driver: MatrixDriver,
+    event_forwarding_guard: Option<DropGuard>,
+    to_widget_tx: Sender<String>,
+    events_tx: UnboundedSender<IncomingMessage>,
+    capabilities_provider: T,
+}
 
-                        MatrixDriverRequestData::SendMatrixEvent(req) => {
-                            let SendEventRequest { event_type, state_key, content } = req;
+impl<T: CapabilitiesProvider> ProcessingContext<T> {
+    async fn process_event(&mut self, event: IncomingMessage) -> Result<(), ()> {
+        for action in self.widget_machine.process(event) {
+            self.process_action(action).await?;
+        }
 
-                            matrix_driver
-                                .send(event_type, state_key, content)
-                                .await
-                                .map(MatrixDriverResponse::MatrixEventSent)
-                                .map_err(|e| e.to_string())
-                        }
+        Ok(())
+    }
+
+    async fn process_action(&mut self, action: Action) -> Result<(), ()> {
+        match action {
+            Action::SendToWidget(msg) => {
+                self.to_widget_tx.send(msg).await.map_err(|_| ())?;
+            }
+            Action::MatrixDriverRequest { request_id, data } => {
+                let response = match data {
+                    MatrixDriverRequestData::AcquireCapabilities(cmd) => {
+                        let obtained = self
+                            .capabilities_provider
+                            .acquire_capabilities(cmd.desired_capabilities.clone())
```

Comment:
Why do we need to clone here? I don't see `cmd` being used again.

Reply:
You're right, an oversight from my side (I just copied this part of code, did not modify it), updated.